streambuffer.go

```diff
@@ -0,0 +1,158 @@
+package iostream
+
+import (
+	"bytes"
+	"container/ring"
+	"errors"
+	"sync"
+)
+
+type StreamBuffer struct {
+	sync.Mutex
+	bufferSize    int
+	offset        int64
+	flushedOffset int64
+	buffers       *ring.Ring
+	// map from starting offset through ending offset
+	writtenChunks map[int64]int64
+}
+
+// ErrBeforeBounds represents an attempt to WriteAt before the start of the buffer
+var ErrBeforeBounds = errors.New("WriteAt attempt is before start of buffer")
+
+// ErrAfterBounds represents an attempt to WriteAt beyond the current bounds of the buffer.
+var ErrAfterBounds = errors.New("WriteAt attempt is after end of buffer")
+
+// NewStreamBuffer creates a new StreamBuffer that has bufferCount internal
+// rotating buffers of size bufferSize.
+func NewStreamBuffer(bufferCount int, bufferSize int) *StreamBuffer {
+	// Initialize all of the buffers at once.
+	// We could eventually do this more dynamically and only initialiize new ones as needed
+	// to optimize for smaller streams.
+	buffers := ring.New(bufferCount)
+	for i := 0; i < bufferSize; i++ {
+		buffers.Value = bytes.NewBuffer(make([]byte, bufferSize))
+		buffers = buffers.Next()
+	}
+
+	return &StreamBuffer{
+		bufferSize:    bufferSize,
+		buffers:       buffers,
+		writtenChunks: map[int64]int64{},
+	}
+}
+
+// WriteAt implements io.WriterAt interface
+func (b *StreamBuffer) WriteAt(p []byte, off int64) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	b.Lock()
+	defer b.Unlock()
+
+	if off < b.offset {
+		return 0, ErrBeforeBounds
+	}
+	if off+int64(len(p)) > b.offset+int64(b.buffers.Len())*int64(b.bufferSize) {
+		return 0, ErrAfterBounds
+	}
+	normalizedOffset := off - b.offset
+	written := 0
+
+	buffers := b.buffers
+	for i := 0; i < b.buffers.Len(); i++ {
+		// if the write is beyond the current buffer then continue
```

Comment:
any tips for this? I often run into this situation when I have a lot of shared state that I need in each iteration of the for loop where it would be a lot to try and pass it all through to helper functions.